tests/cephadm/test_replace_osd.py

```diff
@@ -0,0 +1,85 @@
+from ceph.ceph_admin import CephAdmin
+from ceph.ceph_admin.common import fetch_method
+from ceph.ceph_admin.daemon import Daemon
+from cli.cephadm.cephadm import CephAdm
+from utility.log import Log
+
+log = Log(__name__)
+
+
+def replace_osd(ceph_cluster, osd_id, node_name, base_cmd):
+    installer_node = ceph_cluster.get_nodes(role="installer")[0]
+    cephadm = CephAdm(installer_node)
+    log.info("Replacing OSD {}".format(osd_id))
+    try:
+        cephadm.ceph.orch.osd.rm(osd_id=osd_id, replace=True)
+        log.info("OSD {} has been replaced successfully".format(osd_id))
+    except Exception as e:
+        log.error("Error replacing OSD {}: {}".format(osd_id, e))
+        return 1
+
+
+def add_osd(ceph_cluster, node_name, command, daemon, add_osd_config):
+    log.info("Adding OSD on {}".format(node_name))
+    try:
+        method = fetch_method(daemon, command)
+        method(add_osd_config)
+    except Exception as e:
+        log.error(f"Error adding OSD: {e}")
+        return 1
+
+
+def check_osd_id(osd_id, installer, cephadm):
+    # Get the OSD ID of the new OSD after the replacement
+    cmd = f"ceph osd tree | grep osd.{osd_id} | cut -d' ' -f2"
+    try:
+        cephadm.shell([cmd])
+    except Exception as e:
+        print(f"Failed to execute command {cmd}: {e}")
+        return False
+
+    # Check if OSD is up or not
+    cmd = f"ceph osd dump | grep osd.{osd_id} | awk '{{print $2}}'"
+    try:
+        out = cephadm.shell([cmd])
+    except Exception as e:
+        print(f"Failed to execute command {cmd}: {e}")
+        return False
+
+    if out[0].strip() == "up":
+        return True
+    else:
+        return False
+
+
+def run(ceph_cluster, **kw):
+    config = kw["config"]
+    replace_osd_config = config.get("replace_osd_config", {})
+    add_osd_config = config.get("add_osd_config", {})
+
+    # Replace OSD
+    osd_id_list = replace_osd_config.get("pos_args", [])
+    node_name_list = replace_osd_config.get("nodes", [])
+    base_cmd = replace_osd_config.get("base_cmd_args", [])
+    if not node_name_list or not osd_id_list or not osd_id_list[0]:
+        log.error("Missing node or OSD ID for replace OSD test")
+        return 1
+    osd_id = str(osd_id_list[0])
+    node_name = node_name_list[0]
+    replace_osd(ceph_cluster, osd_id, node_name, base_cmd)
+
+    # Add OSD
+    node_name_list = add_osd_config.get("pos_args", [])
+    if not node_name_list or len(node_name_list) < 2:
+        log.error("Missing node or device for add OSD test")
+        return 1
+    node_name = node_name_list[0]
+    command = add_osd_config.get("command", [])
+    daemon = Daemon(cluster=ceph_cluster, **add_osd_config)
+    add_osd(ceph_cluster, node_name, command, daemon, add_osd_config)
```

Comment:
As these methods are not reused and only perform one step better to use same steps directly here.